### PR TITLE
[Hub Generated] Review request for Microsoft.DocumentDB to add version preview/2020-06-01-preview

### PR DESCRIPTION
**Pull request by tengpanpt**
This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/tengpanpt/azure-rest-api-specs/users%2Ftepa%2Fdev-cosmos-db-Microsoft.DocumentDB-2020-06-01-preview...master/).

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [ ] new service onboarding 
      - [ ] new API version 
      - [x] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [x] PowerShell
      - [x] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [ ] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [ ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [ ] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “**WaitForARMFeedback**” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [ ] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://msazure.visualstudio.com/One/_workitems/create/Scenario?templateId=d1c215ec-5d6a-4cb3-a193-95ddd0a1ba96&ownerId=85bb92fb-21fe-41ca-9309-47f2c5f57515). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json [hunk 1]
```diff
@@ -7787,6 +7826,11 @@
           "format": "int32",
           "minimum": 0,
           "description": "An integer representing the time (in hours) that each backup is retained"
+        },
+        "backupStorageRedundancy": {
+          "description": "Enum to indicate backup residency",
```
Comment (vinhms) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json:7831]: Enum to indicate type of backup residency?

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json [hunk 2]
```diff
@@ -4298,6 +4298,45 @@
...
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.DocumentDB/locations/{location}/backupStorageRedundancies": {
+      "get": {
+        "operationId": "BackupStorageRedundancies_GetByLocation",
+        "x-ms-examples": {
+          "CosmosDBBackupStorageRedundanciesList": {
+            "$ref": "./examples/CosmosDBBackupStorageRedundanciesList.json"
+          }
+        },
+        "description": "Lists all the available backup redundancies in a region.",
```
Comment (vinhms) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json:4310]: Lists all the available backup redundancy types in a region.

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json [hunk 3]
```diff
@@ -7898,6 +7942,48 @@
...
         }
       }
+    },
+    "BackupStorageRedundanciesResult": {
+      "description": "Available backup storage redundancies in selected location.",
+      "type": "object",
+      "properties": {
+        "location": {
+          "type": "string",
+          "description": "The selected location."
```
Comment (vinhms) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json:7952]: please be more descriptive

**Comment by vinhms** [comment]
Added some comments.  Otherwise looks good.

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json (hunk 3 above)
Comment (tengpanpt) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json:7952]: Any suggestions for the description? I have mentioned "Available backup storage redundancy types in selected location." just above so I use "The selected location" here

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json (hunk 2 above)
Comment (tengpanpt) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json]: Thanks updated

**Comment by tengpanpt** [comment]
@erich-wang can you please review this PR and let my know if you have any concern. Thank you! (tepa)

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json [hunk 4]
```diff
@@ -4298,6 +4298,45 @@
           }
         }
       }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.DocumentDB/locations/{location}/backupStorageRedundancies": {
```
Comment (MehaKaushik) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json:4302]: Was this GET api discussed in RP API Review ?

Comment (MehaKaushik) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json:4302]: These URLs represent hierarchical resources. This URL means there is a resource called backupStorageRedundancy under the resource Location. This is not correct. 
This violates both REST API and ARM guidelines. Please re-design this API. Is this API really required? Correct me if I am wrong, this API tells the available values for a region.

**Review by MehaKaushik — CHANGES_REQUESTED** [review]
Please re-design or remove the list backupStorageRedundancies api.

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json (hunk 4 above)
Comment (tengpanpt) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json:4302]: We do have hierarchical resources here. For each location resource, it has it's own set of backupStorageRedundancy.
As for this URL, we once confirmed with Wenbin and Devendra. Initially plan is to expose it to portal, now we also want take this chance to expose it in cli and powershell

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json [hunk 5]
```diff
@@ -4298,6 +4298,45 @@
...
+      "get": {
```
Comment (pilor) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json:4303]: This new resource type needs to have a response body that looks like a resource (id, name, type, etc...) and support an individual GET.  See https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#get-resource

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json [hunk 6]
```diff
@@ -4298,6 +4298,45 @@
...
+          },
+          {
+            "$ref": "../../../../../common-types/resource-management/v1/types.json#/parameters/SubscriptionIdParameter"
+          },
+          {
+            "$ref": "#/parameters/locationParameter"
+          }
+        ],
+        "responses": {
+          "200": {
```
Comment (pilor) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json:4323]: A response to a collection GET needs to look like the paginated response body described here:  https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/resource-api-reference.md#get-resource

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json [hunk 7]
```diff
@@ -4298,6 +4298,45 @@
           }
         }
       }
+    },
```
Comment (pilor) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json:4301]: this new API needs to go in a new api-version

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json (hunk 4 above)
Comment (MehaKaushik) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json:4302]: Looks like Azure storage does similarly https://docs.microsoft.com/en-us/rest/api/storagerp/skus/list
We should be good too.

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json (hunk 5 above)
Comment (tengpanpt) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json]: Removed new API in this PR.

**Review by MehaKaushik — APPROVED** [review]

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json (hunk 6 above)
Comment (tengpanpt) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2020-06-01-preview/cosmos-db.json]: Thanks for the doc, I will not put new API in this PR

**Comment by erich-wang** [comment]
@tengpanpt, adding optional property as considered as breaking change, you need to use new API version unless you're fixing swagger correctness.

NewApiVersionRequired reason:

A service’s API is a contract with customers and is represented by using the api-version query parameter. Changes such as adding an optional property to a request/response or introducing a new operation is a change to the service’s contract and therefore requires a new api-version value. This is critically important for documentation, client libraries, and customer support.

EXAMPLE: if a customer calls a service in the public cloud using api-version=2020-07-27, the new property or operation may exist but if they call the service in a government cloud, air-gapped cloud, or Azure Stack Hub cloud using the same api-version, the property or operation may not exist. Because there is no clear relationship between the service api-version and the new property/operation, customers can’t trust the documentation and Azure customer have difficulty helping customers diagnose issues. In addition, each client library version documents the service version it supports. When an optional property or new operation is added to a service and its Swagger, new client libraries must be produced to expose this functionality to customers. Without updating the api-version, it is unclear to customers which version of a client library supports these new features.

**Comment by tengpanpt** [comment]
> @tengpanpt, adding optional property as considered as breaking change, you need to use new API version unless you're fixing swagger correctness.
> 
> NewApiVersionRequired reason:
> 
> A service’s API is a contract with customers and is represented by using the api-version query parameter. Changes such as adding an optional property to a request/response or introducing a new operation is a change to the service’s contract and therefore requires a new api-version value. This is critically important for documentation, client libraries, and customer support.
> 
> EXAMPLE: if a customer calls a service in the public cloud using api-version=2020-07-27, the new property or operation may exist but if they call the service in a government cloud, air-gapped cloud, or Azure Stack Hub cloud using the same api-version, the property or operation may not exist. Because there is no clear relationship between the service api-version and the new property/operation, customers can’t trust the documentation and Azure customer have difficulty helping customers diagnose issues. In addition, each client library version documents the service version it supports. When an optional property or new operation is added to a service and its Swagger, new client libraries must be produced to expose this functionality to customers. Without updating the api-version, it is unclear to customers which version of a client library supports these new features.

**Comment by tengpanpt** [comment]
> @tengpanpt, adding optional property as considered as breaking change, you need to use new API version unless you're fixing swagger correctness.
> 
> NewApiVersionRequired reason:
> 
> A service’s API is a contract with customers and is represented by using the api-version query parameter. Changes such as adding an optional property to a request/response or introducing a new operation is a change to the service’s contract and therefore requires a new api-version value. This is critically important for documentation, client libraries, and customer support.
> 
> EXAMPLE: if a customer calls a service in the public cloud using api-version=2020-07-27, the new property or operation may exist but if they call the service in a government cloud, air-gapped cloud, or Azure Stack Hub cloud using the same api-version, the property or operation may not exist. Because there is no clear relationship between the service api-version and the new property/operation, customers can’t trust the documentation and Azure customer have difficulty helping customers diagnose issues. In addition, each client library version documents the service version it supports. When an optional property or new operation is added to a service and its Swagger, new client libraries must be produced to expose this functionality to customers. Without updating the api-version, it is unclear to customers which version of a client library supports these new features.

Hi @erich-wang , I am updating the preview version, which is only used for recent new backup and restore features. From  "Breaking Change Review Checklist", it says "Updating API(s) in preview over 1 year". Do I still need new preview version? Thank you!

**Review by JeffreyRichter — APPROVED** [review]

**Pull request merged**